app/src/main/java/ingsw/group1/findmyphone/contacts/SMSContactManager.java

```diff
@@ -0,0 +1,181 @@
+ * because it is assumed that few contacts are inserted and not frequently.
+ * So they can be added and deleted directly from the {@link SMSContactDatabase} every time.
+ * It has been used design pattern Singleton.
+ *
+ * @author Giorgia Bortoletti
+ */
+public class SMSContactManager implements ContactManager<String, SMSPeer, SMSContact>{
+
+    private static final String CONTACTS_DB_NAME = "contact-db";
+
+    private SMSContactDatabase contactDatabase;
+    private static SMSContactManager instance;
+
+    //---------------------------- CONSTRUCTOR ----------------------------
+
+    /**
+     * Private constructor
+     *
+     * @param applicationContext    {@link Context} of the application
+     *
+     * @throws ExceptionInInitializerError if the object is already initialized
+     */
+    private SMSContactManager(@NonNull Context applicationContext) throws ExceptionInInitializerError {
+        if(instance != null)
+            throw new ExceptionInInitializerError();
+
+        contactDatabase = Room.databaseBuilder(applicationContext, SMSContactDatabase.class,
+                CONTACTS_DB_NAME)
+                .enableMultiInstanceInvalidation()
+                .allowMainThreadQueries()
+                .build();
+    }
+
+    /**
+     * Method to get the only valid on-disk instance of this class. A new instance is created only if it was
+     * null previously. The used context is always the parent application context of the parameter.
+     *
+     * @param applicationContext    The calling context
+     *
+     * @return the only instance of {@link SMSContactManager}
+     */
+    public static SMSContactManager getInstance(@NonNull Context applicationContext) {
+        if (instance == null) {
+            instance = new SMSContactManager(applicationContext);
+        }
+        return instance;
+    }
+
+    //---------------------------- OPERATIONS ON THE SINGLE CONTACT ----------------------------
+
+    /**
+     * Verify if a contact could be a valid contact in its phone
+     *
+     * @param contactPhone  Contact phone to verify
+     *
+     * @return true if contact is valid, false otherwise
+     */
+    public boolean isValidContactPhone(@NonNull String contactPhone){
+        try {
+            new SMSPeer(contactPhone);
+        }catch (InvalidTelephoneNumberException addressException){
+            return false;
+        }
+        return true;
+    }
+
+    //---------------------------- OPERATIONS ON THE CONTACTS DATABASE ----------------------------
+
+    /**
+     * Add a {@link SMSPeer} as {@link SMSContact} in {@link SMSContactDatabase}
+     * after using {@link SMSContactConverterUtils} to convert SMSPeer in a Contact entity
+     *
+     * @param peer      {@link SMSPeer} to insert in the contacts database
+     */
+    public void addContact(@NonNull SMSPeer peer) {
+        SMSContact newContact = SMSContactConverterUtils.contactFromPeer(peer);
+        contactDatabase.access().insert(newContact);
```

Comment:
Here you can use method nesting taking advantage of the last method overload, that is `addContact(@NonNull SMSContact newContact)`, by doing this:
```suggestion
        addContact(newContact);
```
Right now there isn't any big advantage (not even a line saved), but if one day, in the future, the process of adding a new contact to the database become more complex then you will have to modify only the last overload instead of all the overloads.

Reply:
You are right.
In other places I used it but here I forgot.